### PR TITLE
Messaging thread cleanup and job support

### PR DESCRIPTION
**Pull request by slagle**
- Run RPC server thread in run_* loop
- Add job handling support for messaging driver

### CODE REVIEW

File: directord/server.py [hunk 1]
```diff
@@ -847,6 +778,86 @@ def handle_heartbeat(self, identity, data):
 
         self.workers[identity] = metadata
 
+    def handle_job(
```
Comment (cloudnull) [inline directord/server.py:781]: this method needs a doc string.

File: directord/server.py [hunk 2]
```diff
@@ -598,92 +598,23 @@ def run_interactions(self, sentinel=False):
                     self.handle_heartbeat(identity, data)
                 else:
                     poller_interval, poller_time = 1, time.time()
-                    self.log.debug(
-                        "Job information received [ %s ] from [ %s ]",
-                        msg_id,
+                    self.handle_job(
```
Comment (cloudnull) [inline directord/server.py:601]: we should use named args.

File: directord/drivers/messaging.py [hunk 3]
```diff
@@ -150,3 +192,21 @@ def send(self, method, topic, server="directord", **kwargs):
     def heartbeat(self, context, identity, data):
         self.log.info("Handling heartbeat")
         self.interface.handle_heartbeat(identity, data)
+
+    @expose
+    def job(
```
Comment (cloudnull) [inline directord/drivers/messaging.py:197]: this method needs a doc string.

File: directord/server.py [hunk 4]
```diff
@@ -586,9 +586,9 @@ def run_interactions(self, sentinel=False):
             if self.driver.job_check(constant=poller_interval):
                 (
                     identity,
-                    msg_id,
```
Comment (cloudnull) [inline directord/server.py:589]: I don't think we should rename this.

File: directord/server.py [hunk 5]
```diff
@@ -598,92 +598,23 @@ def run_interactions(self, sentinel=False):
...
-                                    data=new_task,
-                                )
-                            )
 
             if time.time() > prune_time:
                 self.workers.prune()
                 prune_time = time.time() + 762
 
             if sentinel:
+                self.driver.job_stop()
```
Comment (cloudnull) [inline directord/server.py:617]: the driver spec has job_close, unless we expect messaging to have a specific close method, this should probably just use that. 

https://github.com/cloudnull/directord/blob/main/directord/drivers/__init__.py#L268

File: directord/client.py [hunk 6]
```diff
@@ -801,73 +801,87 @@ def run_job(self, lock=None, sentinel=False):
...
-                        )
+                self.handle_job(command=command, data=data, info=info)
             else:
                 cache_check_time = self.prune_cache(
                     cache_check_time=cache_check_time
                 )
                 time.sleep(poller_interval * 0.001)
 
             if sentinel:
+                self.driver.job_stop()
```
Comment (cloudnull) [inline directord/client.py:812]: this should be `job_close`, unless the messaging driver will use the close method from the abstract elsewhere.

https://github.com/cloudnull/directord/blob/main/directord/drivers/__init__.py#L268

File: directord/drivers/messaging.py [hunk 7]
```diff
@@ -94,7 +94,44 @@ def heartbeat_send(
             data=data,
         )
 
-    def run(self, sentinel=False):
+    def job_send(
```
Comment (cloudnull) [inline directord/drivers/messaging.py:97]: doc string.

File: directord/drivers/messaging.py [hunk 8]
```diff
@@ -117,11 +154,16 @@ def run(self, sentinel=False):
...
         )
         self.log.info("Starting messaging server")
-        server.start()
-        while not sentinel:
-            time.sleep(1)
-        server.stop()
-        server.wait()
+        self.server.start()
+
+    def job_stop(self):
```
Comment (cloudnull) [inline directord/drivers/messaging.py:159]: I think this should be job_close().

**Review by cloudnull — CHANGES_REQUESTED** [review]

File: directord/client.py [hunk 9]
```diff
@@ -801,73 +801,87 @@ def run_job(self, lock=None, sentinel=False):
...
                 break
 
+    def handle_job(
```
Comment (cloudnull) [inline directord/client.py:815]: needs doc string.

File: directord/drivers/__init__.py [hunk 10]
```diff
@@ -281,6 +281,11 @@ def job_init(self):
 
         pass
 
+    def job_stop(self):
```
Comment (cloudnull) [inline directord/drivers/__init__.py:284]: do we need this? I think the message driver should use the existing job_close from the driver abstract.

File: directord/drivers/messaging.py [hunk 11]
```diff
@@ -117,11 +154,16 @@ def run(self, sentinel=False):
...
+        """Stop the server mode."""
+
+        self.log.info("Stopping messaging server")
+        if not self.server:
+            self.log.info("No server to stop")
+        self.server.stop()
```
Comment (cloudnull) [inline directord/drivers/messaging.py:165]: these should probable be within an `else`

File: directord/server.py [hunk 12]
```diff
@@ -847,6 +778,86 @@ def handle_heartbeat(self, identity, data):
...
+            "Job information received [ %s ] from [ %s ]",
+            job_id,
+            identity,
+        )
+        node = identity
+        node_output = info
+        if stderr:
+            stderr = stderr
+        if stdout:
+            stdout = stdout
```
Comment (cloudnull) [inline directord/server.py:794]: this is a remnant from encoding, we can probably drop it.

File: directord/server.py [hunk 13]
```diff
@@ -847,6 +778,86 @@ def handle_heartbeat(self, identity, data):
...
+    ):
+        self.log.debug(
+            "Job information received [ %s ] from [ %s ]",
+            job_id,
+            identity,
+        )
+        node = identity
+        node_output = info
+        if stderr:
+            stderr = stderr
```
Comment (cloudnull) [inline directord/server.py:792]: this is a remnant from encoding, we can probably drop it.

File: directord/server.py [hunk 14]
```diff
@@ -847,6 +778,86 @@ def handle_heartbeat(self, identity, data):
...
 
+    def handle_job(
+        self, identity, job_id, control, command, data, info, stderr, stdout
+    ):
+        self.log.debug(
+            "Job information received [ %s ] from [ %s ]",
+            job_id,
+            identity,
+        )
+        node = identity
```
Comment (cloudnull) [inline directord/server.py:789]: this is a remnant from encoding, we can probably drop it.

File: directord/server.py (hunk 1 above)
Comment (slagle) [reply directord/server.py:781]: Done.

File: directord/server.py (hunk 2 above)
Comment (slagle) [reply directord/server.py:601]: Done.

File: directord/drivers/messaging.py (hunk 3 above)
Comment (slagle) [reply directord/drivers/messaging.py:197]: Done.

File: directord/server.py (hunk 4 above)
Comment (slagle) [reply directord/server.py:589]: Restored

File: directord/server.py (hunk 5 above)
Comment (slagle) [reply directord/server.py:617]: Done.

File: directord/client.py (hunk 6 above)
Comment (slagle) [reply directord/client.py:812]: Done.

File: directord/drivers/messaging.py (hunk 8 above)
Comment (slagle) [reply directord/drivers/messaging.py:159]: Done.

File: directord/drivers/messaging.py (hunk 7 above)
Comment (slagle) [reply directord/drivers/messaging.py:97]: Done.

**Pull request merged**